### PR TITLE
[fix] checking for compatible redis-server version. 

### LINKED ISSUE #392: RedisTimeSeries Module Crashes ON LOAD
Copy & paste of the BUG REPORT below...

**Module Location & Permissions:**
```
$ ls -l /etc/redis
-rw-r----- 1 redis redis  58884 May  8 17:26 redis.conf
-rwxrwxrwx 1 redis redis 146824 May  8 17:36 redistimeseries.so
```

**Redis Command Executed:**
```
redis-cli -h localhost -p 6379 -a $REDIS_PASS
localhost:6379> MODULE LOAD /etc/redis/redistimeseries.so
Error: Connection reset by peer
```

**Bug Report**
```
=== REDIS BUG REPORT START: Cut & paste starting from here ===
18371:M 08 May 17:53:57.003 # Redis 4.0.9 crashed by signal: 11
18371:M 08 May 17:53:57.003 # Accessing address: (nil)
18371:M 08 May 17:53:57.003 # Failed assertion: <no assertion failed> (<no file>:0)

------ STACK TRACE ------
/usr/bin/redis-server 0.0.0.0:6379(logStackTrace+0x5a)[0x55edb0dd4f1a]
/usr/bin/redis-server 0.0.0.0:6379(sigsegvHandler+0xb1)[0x55edb0dd56d1]
/lib/x86_64-linux-gnu/libpthread.so.0(+0x12890)[0x7f54bcfc0890]

------ INFO OUTPUT ------
# Server
redis_version:4.0.9
redis_git_sha1:00000000
redis_git_dirty:0
redis_build_id:9435c3c2879311f3
redis_mode:standalone
os:Linux 4.15.0-99-generic x86_64
arch_bits:64
multiplexing_api:epoll
atomicvar_api:atomic-builtin
gcc_version:7.4.0
process_id:18371
run_id:e6d617803f1d17acbab2fea8ecbd696eb8e140ea
tcp_port:6379
uptime_in_seconds:61
uptime_in_days:0
hz:10
lru_clock:11902132
executable:/usr/bin/redis-server
config_file:/etc/redis/redis.conf

# Clients
connected_clients:1
client_longest_output_list:0
client_biggest_input_buf:0
blocked_clients:0

# Memory
used_memory:841672
used_memory_human:821.95K
used_memory_rss:3813376
used_memory_rss_human:3.64M
used_memory_peak:841672
used_memory_peak_human:821.95K
used_memory_peak_perc:100.16%
used_memory_overhead:832190
used_memory_startup:782560
used_memory_dataset:9482
used_memory_dataset_perc:16.04%
total_system_memory:2089897984
total_system_memory_human:1.95G
used_memory_lua:37888
used_memory_lua_human:37.00K
maxmemory:0
maxmemory_human:0B
maxmemory_policy:noeviction
mem_fragmentation_ratio:4.53
mem_allocator:jemalloc-3.6.0
active_defrag_running:0
lazyfree_pending_objects:0

# Persistence
loading:0
rdb_changes_since_last_save:0
rdb_bgsave_in_progress:0
rdb_last_save_time:1588960375
rdb_last_bgsave_status:ok
rdb_last_bgsave_time_sec:-1
rdb_current_bgsave_time_sec:-1
rdb_last_cow_size:0
aof_enabled:0
aof_rewrite_in_progress:0
aof_rewrite_scheduled:0
aof_last_rewrite_time_sec:-1
aof_current_rewrite_time_sec:-1
aof_last_bgrewrite_status:ok
aof_last_write_status:ok
aof_last_cow_size:0

# Stats
total_connections_received:1
total_commands_processed:2
instantaneous_ops_per_sec:0
total_net_input_bytes:154
total_net_output_bytes:10168
instantaneous_input_kbps:0.00
instantaneous_output_kbps:0.00
rejected_connections:0
sync_full:0
sync_partial_ok:0
sync_partial_err:0
expired_keys:0
expired_stale_perc:0.00
expired_time_cap_reached_count:0
evicted_keys:0
keyspace_hits:0
keyspace_misses:0
pubsub_channels:0
pubsub_patterns:0
latest_fork_usec:0
migrate_cached_sockets:0
slave_expires_tracked_keys:0
active_defrag_hits:0
active_defrag_misses:0
active_defrag_key_hits:0
active_defrag_key_misses:0

# Replication
role:master
connected_slaves:0
master_replid:94cd0e1d9656926bc3e2aabfc298643ce636161e
master_replid2:0000000000000000000000000000000000000000
master_repl_offset:0
second_repl_offset:-1
repl_backlog_active:0
repl_backlog_size:1048576
repl_backlog_first_byte_offset:0
repl_backlog_histlen:0

# CPU
used_cpu_sys:0.04
used_cpu_user:0.02
used_cpu_sys_children:0.00
used_cpu_user_children:0.00

# Commandstats
cmdstat_auth:calls=1,usec=2,usec_per_call=2.00
cmdstat_command:calls=1,usec=353,usec_per_call=353.00

# Cluster
cluster_enabled:0

# Keyspace

------ CLIENT LIST OUTPUT ------
id=3 addr=127.0.0.1:58258 fd=8 name= age=9 idle=0 flags=N db=0 sub=0 psub=0 multi=-1 qbuf=0 qbuf-free=32768 obl=0 oll=0 omem=0 events=r cmd=module

------ CURRENT CLIENT INFO ------
id=3 addr=127.0.0.1:58258 fd=8 name= age=9 idle=0 flags=N db=0 sub=0 psub=0 multi=-1 qbuf=0 qbuf-free=32768 obl=0 oll=0 omem=0 events=r cmd=module
argv[0]: 'MODULE'
argv[1]: 'LOAD'
argv[2]: '/etc/redis/redistimeseries.so'

------ REGISTERS ------
18371:M 08 May 17:53:57.005 #
RAX:00007f54ba7fc798 RBX:00007f54ba7fca28
RCX:0000000000000006 RDX:00007f54ba7fcc10
RDI:0000000000000000 RSI:00007f54bc010c90
RBP:00007ffd82e33730 RSP:00007ffd82e33698
R8 :0000000000000000 R9 :0000000000000000
R10:0000000000000000 R11:00007f54ba5f7810
R12:00007f54ba7fca90 R13:00007f54ba7fc750
R14:0000000000000000 R15:0000000000000000
RIP:0000000000000000 EFL:0000000000010246
CSGSFS:002b000000000033
18371:M 08 May 17:53:57.005 # (00007ffd82e336a7) -> 0000000000000000
18371:M 08 May 17:53:57.005 # (00007ffd82e336a6) -> 0000000000000000
18371:M 08 May 17:53:57.005 # (00007ffd82e336a5) -> 00007f54bc0ed393
18371:M 08 May 17:53:57.005 # (00007ffd82e336a4) -> 00007f54bc091a00
18371:M 08 May 17:53:57.005 # (00007ffd82e336a3) -> 000055edb0e01735
18371:M 08 May 17:53:57.005 # (00007ffd82e336a2) -> 6460a5f0f8d05100
18371:M 08 May 17:53:57.005 # (00007ffd82e336a1) -> 00007f54ba5f59d0
18371:M 08 May 17:53:57.005 # (00007ffd82e336a0) -> 0000000000000000
18371:M 08 May 17:53:57.005 # (00007ffd82e3369f) -> 00007f54ba5f5b70
18371:M 08 May 17:53:57.005 # (00007ffd82e3369e) -> 00007f54ba5f7220
18371:M 08 May 17:53:57.005 # (00007ffd82e3369d) -> 00007f54ba5f5560
18371:M 08 May 17:53:57.005 # (00007ffd82e3369c) -> 00007f54ba5f52b0
18371:M 08 May 17:53:57.005 # (00007ffd82e3369b) -> 0000000000000001
18371:M 08 May 17:53:57.005 # (00007ffd82e3369a) -> 00007f54ba5f4c5a
18371:M 08 May 17:53:57.006 # (00007ffd82e33699) -> 0000000000000000
18371:M 08 May 17:53:57.006 # (00007ffd82e33698) -> 00007f54ba5f083f

------ FAST MEMORY TEST ------
18371:M 08 May 17:53:57.006 # Bio thread for job type #0 terminated
18371:M 08 May 17:53:57.006 # Bio thread for job type #1 terminated
18371:M 08 May 17:53:57.006 # Bio thread for job type #2 terminated
*** Preparing to test memory region 55edb106f000 (90112 bytes)
*** Preparing to test memory region 7f54ba7fe000 (8388608 bytes)
*** Preparing to test memory region 7f54bafff000 (8388608 bytes)
*** Preparing to test memory region 7f54bb800000 (8388608 bytes)
*** Preparing to test memory region 7f54bc000000 (8388608 bytes)
*** Preparing to test memory region 7f54bcfaa000 (16384 bytes)
*** Preparing to test memory region 7f54bd1c9000 (16384 bytes)
*** Preparing to test memory region 7f54bd402000 (4096 bytes)
*** Preparing to test memory region 7f54bdbc1000 (16384 bytes)
*** Preparing to test memory region 7f54bdbce000 (4096 bytes)
.O.O.O.O.O.O.O.O.O.O
Fast memory test PASSED, however your memory can still be broken. Please run a memory test for several hours if possible.

=== REDIS BUG REPORT END. Make sure to include from START to END. ===
```

### LINKED ISSUE #393: redis-server version should be verified on load
#226 could be avoided by this 
#392 as well 
#294 also crashed ( 4.0.11 )

### LINKED ISSUE #394: redistimeseries git_sha should be logged on load

### PR DESCRIPTION
**Pull request by filipecosta90**
- logging redistimeseries module version, and git_sha.
- checking for compatible redis-server version ( at least 5.0.0 given the RedisModuleDict ) cc @K-Jo 
- using same code as RedisGears to check for redis-server version and setup

fixes #392 
fixes #393 
fixes #394

### CODE REVIEW

**Comment by ashtul** [comment]
@filipecosta90 
Maybe you can try and get this into RedisModule API as well? It will lack back compatibility but will be useful in the future.

File: src/config.c [hunk 1]
```diff
@@ -52,4 +57,81 @@ int ReadConfig(RedisModuleCtx *ctx, RedisModuleString **argv, int argc) {
     }
     RedisModule_Log(ctx, "verbose", "loaded default MAX_SAMPLE_PER_CHUNK policy: %lld \n", TSGlobalConfig.maxSamplesPerChunk);
     return TSDB_OK;
+}
+
+RedisVersion currVersion;
```
Comment (MeirShpilraien) [inline src/config.c:62]: Namespace

File: src/config.c [hunk 2]
```diff
@@ -52,4 +57,81 @@ int ReadConfig(RedisModuleCtx *ctx, RedisModuleString **argv, int argc) {
...
+
+RedisVersion supportedVersion = {
```
Comment (MeirShpilraien) [inline src/config.c:64]: Namespace

File: src/config.h [hunk 3]
```diff
@@ -21,4 +22,25 @@ typedef struct {
 extern TSConfig TSGlobalConfig;
 
 int ReadConfig(RedisModuleCtx *ctx, RedisModuleString **argv, int argc);
+
+typedef struct RedisVersion {
```
Comment (MeirShpilraien) [inline src/config.h:26]: Namespace

File: src/config.h [hunk 4]
```diff
@@ -21,4 +22,25 @@ typedef struct {
...
+extern RedisVersion supportedVersion;
+
+extern int timeseriesRlecMajorVersion;
+extern int timeseriesRlecMinorVersion;
+extern int timeseriesRlecPatchVersion;
+extern int timeseriesRlecBuild;
+
+extern bool timeseriesIsCrdt;
+
+static inline int IsEnterprise() { return timeseriesRlecMajorVersion != -1; }
```
Comment (MeirShpilraien) [inline src/config.h:42]: Namespace

**Comment by filipecosta90** [comment]
@ashtul @MeirShpilraien can you guys please revise. prefixed all functions and vars with RTS_*.

File: src/config.c [hunk 5]
```diff
@@ -8,6 +8,11 @@
 #include "rmutil/strings.h"
 #include "config.h"
 #include "consts.h"
+#include "common.h"
+#include <string.h>
+#include <stdbool.h>
+#include <stdarg.h>
+#include <assert.h>
```
Comment (danni-m) [inline src/config.c:15]: @filipecosta90 do we need all of these includes?

File: src/config.c [hunk 6]
```diff
@@ -52,4 +57,69 @@ int ReadConfig(RedisModuleCtx *ctx, RedisModuleString **argv, int argc) {
     }
     RedisModule_Log(ctx, "verbose", "loaded default MAX_SAMPLE_PER_CHUNK policy: %lld \n", TSGlobalConfig.maxSamplesPerChunk);
     return TSDB_OK;
+}
+
+RTS_RedisVersion RTS_currVersion;
+
+RTS_RedisVersion RTS_supportedVersion = {
```
Comment (danni-m) [inline src/config.c:64]: why not name it `RTS_minSupportedVersion`?

File: src/config.c (hunk 5 above)
Comment (filipecosta90) [reply src/config.c:15]: string.h - strstr
assert.h - assert 
stdbool.h ( not needed after removal of crdt check )
stdarg ( not needed )
removing :)

File: src/config.c (hunk 6 above)
Comment (filipecosta90) [reply src/config.c]: agree. changed in new revision.

**Comment by K-Jo** [comment]
@filipecosta90 shouldn't it be in the ramp.yaml file as well?

**Comment by filipecosta90** [comment]
> @filipecosta90 shouldn't it be in the ramp.yaml file as well?

as @danni-m stated, already there :)
https://github.com/RedisTimeSeries/RedisTimeSeries/blob/master/ramp.yml#L8

**Pull request merged**